pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,651 @@
+			}
+		}
+	}
+
+	c.runningPacketSamplingsMutex.Lock()
+	psState := packetSamplingState{
+		name: ps.Name, tag: ps.Status.DataplaneTag,
+		receiverOnly: receiverOnly, isSender: isSender,
+		maxNumCapturedPackets: ps.Spec.FirstNSamplingConfig.Number,
+	}
+
+	exists, err := fileExists(string(ps.UID))
+	if err != nil {
+		return fmt.Errorf("couldn't check if the file exists: %w", err)
+
+	}
+	if exists {
+		return fmt.Errorf("packet file already exists. this may be due to an unexpected termination")
+	}
+
+	file, err := createPcapngFile(string(ps.UID))
+	if err != nil {
+		return fmt.Errorf("couldn't craete pcapng file: %w", err)
+	}
+
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return fmt.Errorf("couldn't init pcap writer: %w", err)
+	}
+
+	psState.shouldSyncPackets = len(podInterfaces) > 0
+	psState.uid = string(ps.UID)
+	psState.pcapngFile = file
+	psState.pcapngWriter = writer
+
+	if psState.shouldSyncPackets {
+		psState.updateRateLimiter = rate.NewLimiter(rate.Every(samplingStatusUpdatePeriod), 1)
+	}
+	c.runningPacketSamplings[psState.tag] = &psState
+	c.runningPacketSamplingsMutex.Unlock()
+
+	timeout := ps.Spec.Timeout
+	if timeout == 0 {
+		timeout = crdv1alpha1.DefaultPacketSamplingTimeout
+	}
+	if psState.shouldSyncPackets {
+		klog.V(2).InfoS("installing flow entries for PacketSampling.", "name", ps.Name)
+		err = c.ofClient.InstallPacketSamplingFlows(uint8(psState.tag), senderOnly, receiverOnly, matchPacket, endpointPackets, ofPort, timeout)
+		if err != nil {
+			klog.ErrorS(err, "install flow entries failed.", "name", ps.Name)
+		}
+	}
+	return err
+
+}
+
+func createPcapngFile(uid string) (afero.File, error) {
+	return defaultFS.Create(uidToPath(uid))
+}
+
+func fileExists(uid string) (bool, error) {
+	_, err := defaultFS.Stat(uidToPath(uid))
+	if err == nil {
+		return true, nil
+	} else {
+		if os.IsNotExist(err) {
+			return false, nil
+		} else {
+			return false, err
+		}
+	}
+}
+
+// genEndpointMatchPackets generates match packets(with dest endpoint's ip/port info) besides the normal match packet.
+// these match packets will help the pipeline to capture the pod -> svc traffic.
+// TODO: 1. support name based port name 2. dual-stack support
+func (c *Controller) genEndpointMatchPackets(ps *crdv1alpha1.PacketSampling) ([]binding.Packet, error) {
+	if ps.Spec.Destination.Service != "" {
+		var port int32
+		if ps.Spec.Packet.TransportHeader.TCP != nil {
+			port = ps.Spec.Packet.TransportHeader.TCP.DstPort
+		} else if ps.Spec.Packet.TransportHeader.UDP != nil {
+			port = ps.Spec.Packet.TransportHeader.UDP.DstPort
+		}
+
+		var packets []binding.Packet
+		dstSvc, err := c.serviceLister.Services(ps.Spec.Destination.Namespace).Get(ps.Spec.Destination.Service)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, item := range dstSvc.Spec.Ports {
+			if item.Port == port {
+				if item.TargetPort.Type == intstr.Int {
+					port = item.TargetPort.IntVal
+				}
+			}
+		}
+
```

Comment:
I think you need to check the port matching and return error if there is no port matched? otherwise, I didn't see there is any point to check the port here.

Reply:
this was for the case of container port is no the same as service port. Say service port is 80 but the container port is 8080, for the endpoints list, we need to use the containerPort.